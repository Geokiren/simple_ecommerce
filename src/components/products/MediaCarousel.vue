<template>
  <div class="media-slider flex flex-col gap-1 w-full">
    <div class="selected-media w-full">
      <div class="relative aspect-video bg-secondary rounded-lg">
        <OptimizedImage 
          v-if="activeImage"
          :key="activeImage"
          class="w-full h-full object-cover"  
          :src="activeImage" 
          alt="Product Image" 
          @load="onImageLoad" 
        />
      </div>
    </div>

    <div v-if="areImagesGreaterThanOne" class="thumbnails flex gap-1 w-full h-16 lg:h-24 4xl:h-32 overflow-auto">
      <OptimizedImage 
        v-for="(image, index) in images" 
        :key="image" 
        class="cursor-pointer rounded-lg  bg-secondary" 
        :class="{ 'border-2 border-active': image === activeImage }" 
        :src="image"
        :alt="`thumbnail-${index}`"
        loading="lazy"
        @click="activeImage = image" 
      /> 
    </div>
  </div>
</template>


<script setup lang="ts">
  import { ref, computed, defineProps, onMounted, onUpdated } from 'vue';

  import OptimizedImage from '../shared/OptimizedImage.vue';

  interface Props {
    images: string[];
  }

  const { images } = defineProps<Props>();

  const imageLoaded = ref<boolean>(false);

  const onImageLoad = () => {
    imageLoaded.value = true;
  }

  const activeImage = ref('');

  const areImagesGreaterThanOne = computed(() => {
    return images.length > 1;
  });

  onMounted(() => {
    activeImage.value = images[0];
  });

  onUpdated(() => {
    activeImage.value = images[0];
  });
</script>
