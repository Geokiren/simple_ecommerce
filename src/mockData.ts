export const products = [
  {
    id: 1,
    title: 'Product 1',
    description: 'Product 1 description',
    category: 'shoe',
    price: 10.99,
    discountPercentage: 10,
    rating: 4.5,
    stock: 0,
    tags: [],
    brand: '',
    sku: '',
    weight: 0,
    dimensions: {
      width: 0,
      height: 0,
      depth: 0
    },
    warrantyInformation: '',
    shippingInformation: '',
    availabilityStatus: '',
    reviews: [
      {
        reviewerName: 'John Doe',
        reviewerEmail: 'vqzJZ@example.com',
        rating: 5,
        comment: 'Great product!',
        date: '2022-01-01'
      }
    ],
    returnPolicy: '',
    minimumOrderQuantity: 0,
    meta: {
      createdAt: '',
      updatedAt: '',
      barcode: '',
      qrCode: ''
    },
    thumbnail: '',
    images: [],
    quantity: 1
  },
  {
    id: 2,
    title: 'Product 2',
    description: '',
    category: '',
    price: 20.49,
    discountPercentage: 0,
    rating: 0,
    stock: 0,
    tags: [],
    brand: '',
    sku: '',
    weight: 0,
    dimensions: {
      width: 0,
      height: 0,
      depth: 0
    },
    warrantyInformation: '',
    shippingInformation: '',
    availabilityStatus: '',
    reviews: [],
    returnPolicy: '',
    minimumOrderQuantity: 0,
    meta: {
      createdAt: '',
      updatedAt: '',
      barcode: '',
      qrCode: ''
    },
    thumbnail: '',
    images: [],
    quantity: 1
  }
];
