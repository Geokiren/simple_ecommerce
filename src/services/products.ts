const apiBaseUrl: string = `${import.meta.env.VITE_API_BASE_URL as string}/products`;

export const getCategories = async () => {
  try {
    const response = await fetch(`${apiBaseUrl}/categories`);

    if (!response.ok) {
      throw new Error(`Failed to fetch product categories. HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const searchProducts = async (searchTerm: string) => {
  try {
    const response = await fetch(`${apiBaseUrl}/search?q=${searchTerm}`);

    if (!response.ok) {
      throw new Error(`Failed to fetch products. HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const getProducts = async (params: string) => {
  try {
    const response = await fetch(`${apiBaseUrl}?${new URLSearchParams(params).toString()}`);

    if (!response.ok) {
      throw new Error(`Failed to fetch products. HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const getProductsByCategory = async (category: string, params: string) => {
  try {
    const response = await fetch(
      `${apiBaseUrl}/category/${category}?${new URLSearchParams(params).toString()}`
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch products. HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};

export const getProduct = async (id: string) => {
  try {
    const response = await fetch(`${apiBaseUrl}/${id}`);

    if (!response.ok) {
      throw new Error(`Failed to fetch product. HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error(error);
    throw error;
  }
};
