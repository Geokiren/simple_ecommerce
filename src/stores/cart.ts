import { ref, computed } from 'vue';
import { defineStore } from 'pinia';

import type { Product } from '../types';

export const useCartStore = defineStore('cart', () => {
  const cartProducts = ref<Product[]>([]);

  const cartCounter = computed(() =>
    cartProducts.value.reduce((acc, curr) => acc + (curr.quantity ?? 0), 0)
  );

  const totalPrice = computed(() =>
    parseFloat(
      cartProducts.value
        .reduce((acc, curr) => acc + curr.price * (curr.quantity ?? 0), 0)
        .toFixed(2)
    )
  );

  const addToBasket = (product: Product) => {
    const foundProduct = cartProducts.value.find((p) => p.id === product.id);

    if (!foundProduct) {
      const newProduct = { ...product, quantity: 1 };
      cartProducts.value.push(newProduct);
    } else {
      foundProduct.quantity = (foundProduct.quantity ?? 0) + 1;
    }

    cartProducts.value = [...cartProducts.value];
  };

  const removeFromBasket = (product: Product) => {
    cartProducts.value = cartProducts.value.filter((p) => p.id !== product.id);
  };

  return { cartProducts, cartCounter, totalPrice, addToBasket, removeFromBasket };
});
