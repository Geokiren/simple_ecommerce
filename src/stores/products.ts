import { ref, computed, watch } from 'vue';
import { defineStore } from 'pinia';

import {
  getCategories,
  searchProducts,
  getProducts,
  getProductsByCategory,
  getProduct
} from '@/services/products';

import type { Product } from '../types';
import type { Category } from '../types/category';

interface IPaging {
  limit: number;
  skip: number;
}

interface ISortBy {
  sortBy: string;
  order: string;
}

export const useProductsStore = defineStore('products', () => {
  const categories = ref<Category[]>([]);
  const products = ref<Product[]>([]);

  const paging = ref<IPaging>({
    limit: 30,
    skip: 0
  });

  const total = ref(0);

  const searchTerm = ref<string>('');

  const sorting = ref<ISortBy>({
    sortBy: 'title',
    order: 'asc'
  });

  const activeCategory = ref<Category | null>(null);

  const loading = ref(false);

  const filters = computed(() => {
    const sortBy = sorting.value.sortBy ? `&sortBy=${sorting.value.sortBy}` : '';
    const order = sortBy ? `&order=${sorting.value.order}` : '';
    const pagingParams = `&limit=${paging.value.limit}&skip=${paging.value.skip}`;

    return `?${new URLSearchParams(sortBy + order + pagingParams).toString()}`;
  });

  const setSearchTerm = (term: string) => {
    searchTerm.value = term;
  };

  const setActiveCategory = (category: Category | null) => {
    activeCategory.value = category;
  };

  const setSorting = (value: string, property: 'sortBy' | 'order') => {
    sorting.value[property] = value;
  };

  const hasMorePages = computed(() => {
    return products.value.length < total.value;
  });

  const incrementPage = () => {
    if (hasMorePages.value) {
      paging.value = { ...paging.value, skip: paging.value.skip + paging.value.limit };
    }
  };

  function debounce<T extends (...args: any[]) => any>(fn: T, delay: number) {
    let timeoutId: ReturnType<typeof setTimeout> | null = null;

    return function (this: any, ...args: Parameters<T>) {
      if (timeoutId !== null) {
        clearTimeout(timeoutId);
      }

      timeoutId = setTimeout(() => {
        fn.apply(this, args);
        timeoutId = null;
      }, delay);
    };
  }

  const debouncedSearch = debounce(async () => {
    await fetchSearchResults();
  }, 300);

  watch(searchTerm, async (value: string) => {
    if (!value) {
      products.value = [];
      fetchProducts();
      return;
    }

    if (value.length >= 3) {
      products.value = [];
      debouncedSearch();
    }
  });

  watch(activeCategory, (value: Category | null) => {
    if (!value) {
      products.value = [];
      fetchProducts();
      return;
    }

    products.value = [];
    fetchProductsByCategory();
  });

  watch(paging, (newValue, oldValue) => {
    if (activeCategory.value) {
      if (newValue.limit !== oldValue.limit) products.value = [];
      fetchProductsByCategory();
      return;
    }

    if (newValue.limit !== oldValue.limit) products.value = [];
    fetchProducts();
  });

  watch(
    sorting,
    () => {
      if (activeCategory.value) {
        products.value = [];
        fetchProductsByCategory();
      } else {
        products.value = [];
        fetchProducts();
      }
    },
    { deep: true }
  );

  const fetchCategories = async (): Promise<Category[]> => {
    try {
      const categories: Category[] = await getCategories();
      return categories;
    } catch (error) {
      console.error(error);
      throw error;
    }
  };

  const fetchSearchResults = async (): Promise<{ products: Product[]; total: number }> => {
    loading.value = true;

    try {
      const data = await searchProducts(searchTerm.value);

      products.value = data.products;
      total.value = data.total;

      return { products: data.products, total: data.total };
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      loading.value = false;
    }
  };

  const fetchProducts = async (): Promise<{ products: Product[]; total: number }> => {
    loading.value = true;

    try {
      const data = await getProducts(filters.value);

      products.value = [...products.value, ...data.products];
      total.value = data.total;

      return { products: data.products, total: data.total };
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      loading.value = false;
    }
  };

  const fetchProductsByCategory = async (returnData?: boolean, category?: string) => {
    if (!category && !activeCategory.value) {
      console.error('No category provided');
      return;
    }

    loading.value = true;

    try {
      const categoryToUse = category || activeCategory.value?.slug || '';

      const data = await getProductsByCategory(categoryToUse, filters.value);

      if (returnData) {
        return data?.products || [];
      } else {
        products.value = [...products.value, ...data.products];
        total.value = data.total;
      }
    } catch (error) {
      console.error(error);
    } finally {
      loading.value = false;
    }
  };

  const fetchProduct = async (id: string) => {
    try {
      const product = await getProduct(id);
      return product;
    } catch (error) {
      console.error(error);
    }
  };

  return {
    categories,
    activeCategory,
    products,
    searchTerm,
    loading,
    fetchCategories,
    fetchSearchResults,
    fetchProductsByCategory,
    fetchProducts,
    fetchProduct,
    setSearchTerm,
    setActiveCategory,
    setSorting,
    incrementPage
  };
});
