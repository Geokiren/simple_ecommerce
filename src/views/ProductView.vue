<template>
  <section v-if="isProductValid" class="max-w-screen-2xl mx-auto px-6 py-6 flex flex-col gap-4" data-test="product-view">
    <div class="flex flex-col lg:flex-row">
      <div class="w-full lg:w-2/3 h-auto lg:pr-2">
        <MediaCarousel v-if="product?.images?.length" :images="product?.images || []" />
      </div>

      <div class="w-full lg:w-1/3 mt-6 lg:mt-0 lg:ps-2 flex flex-col gap-2 lg:gap-10">
        <h1 class="text-xl text-center font-bold mb-2" data-test="product-title">{{ product?.title }}</h1>

        <div class="flex justify-between items-center">
          <div class="flex gap-1 flex-wrap">
            <ProductTag v-for="tag in product?.tags" :key="tag" :tag="tag" />
          </div>

          <div class="flex gap-1 items-center">
            <span data-test="product-rating">{{ product?.rating }}</span>
            <IconStar classes="size-5" />
          </div>
        </div>

        <p class="text-sm leading-6 font-light" data-test="product-description">{{ product?.description }}</p>

        <div class="flex">
          <div v-if="product?.discountPercentage" class="p-2 bg-primary text-xl font-semibold rounded-s text-alttext dark:text-text" data-test="product-discount">{{ product?.discountPercentage }}%</div>
          <div class="text-xl font-semibold rounded p-2 bg-secondary rounded-e" data-test="product-price">â‚¬{{ product?.price }}</div>
        </div>

        <div class="flex flex-col gap-2">
          <button 
            class="btn bg-primary rounded p-4 text-lg hover:bg-active dark:hover:text-alttext transition duration-300 ease-in-out text-alttext dark:text-text mt-4 lg:mt-0"
            data-test="product-add-to-bag"
            @click="addToBasket"
          >
            ADD TO BAG
          </button>

          <div class="text-xs lg:text-sm text-center">
            <span data-test="product-return-policy">{{ product?.returnPolicy }}</span>
            <span> | </span>
            <span data-test="product-shipping-info">{{ product?.shippingInformation }}</span>
            <span> | </span>
            <span data-test="product-warranty-info">{{ product?.warrantyInformation }}</span>
          </div>
        </div>
      </div>
    </div>

    <div v-if="categoryProducts.length">
      <h1 class="text-xl text-center lg:text-left font-bold px-4 pt-4 pb-2 w-full border-b">Products in category</h1>

      <GridContainer>
        <ProductCard 
          v-for="product in filteredCategoryProducts" 
          :key="product.id" 
          :product="product"
          :isLast="false"
        />
      </GridContainer>
    </div>
 
    <div class="flex flex-col gap-2 lg:gap-4">
      <h1 class="text-xl text-center lg:text-left font-bold px-4 pt-4 pb-2 w-full border-b">Reviews</h1>

      <div class="grid grid-cols-1 lg:grid-cols-3 gap-4">
        <div class="p-4 bg-secondary rounded-lg shadow-md" v-for="review in reviews" :key="review.id">
          <div class="flex items-center justify-between">
            <div class="flex items-center gap-2">
              <div class="w-12 h-12 rounded-full overflow-hidden">
                <div class="rounded-full w-full h-full bg-primary text-alttext dark:text-text font-bold text-lg flex items-center justify-center">
                  <div data-test="review-initials">{{ review.reviewerName.split(' ').map(name => name[0]).join('') }}</div>             
                </div>
              </div>
              <div class="text-lg font-bold" data-test="review-name">{{ review.reviewerName }}</div>
            </div>
            <div class="flex items-center gap-1">
              <span class="text-lg" data-test="review-rating">{{ review.rating }}</span>
              <IconStar classes="size-5" />
            </div>
          </div>
          
          <div class="mt-2 text-lg">
            <p data-test="review-comment">{{ review.comment }}</p>
          </div>
        </div>
      </div>
    </div>
  </section>

  <InvalidView v-else />
</template>

<script setup lang="ts">
  import { ref, onMounted, computed, nextTick, watch } from 'vue';
  import { useRoute } from 'vue-router';
  import { useProductsStore } from '@/stores/products';
  import { useCartStore } from '@/stores/cart';
  import { storeToRefs } from 'pinia';
  import type { Product } from '../types';

  import MediaCarousel from '@/components/products/MediaCarousel.vue';
  import ProductTag from '@/components/products/ProductTag.vue';
  import IconStar from '@/components/icons/IconStar.vue';
  import GridContainer from '@/layout/GridContainer.vue';
  import ProductCard from '@/components/products/ProductCard.vue';
  import InvalidView from './InvalidView.vue';

  const route = useRoute();

  const productsStore = useProductsStore();
  const { products } = storeToRefs(productsStore);

  const cartStore = useCartStore();

  const productId = ref(route.params.id);
  const product = ref<Product | undefined>({} as Product);
  const categoryProducts = ref<Product[]>([]);
  const isProductValid = ref(true);

  const reviews = computed(() => product.value?.reviews?.map(review => ({ ...review, id: crypto.randomUUID() })) || []);
  const filteredCategoryProducts = computed(() => categoryProducts.value.filter(product => product.id !== Number(productId.value)));

  const getProduct = async () => {
    const id = Array.isArray(productId.value) ? productId.value[0] : productId.value;
    return await productsStore.fetchProduct(id);
  };

  const addToBasket = () => {
    if (!product.value) return;

    cartStore.addToBasket(product.value);
  };

  watch(
    () => route.params.id,
    async (newId) => {
      productId.value = newId;
      product.value = await getProduct();
    },
  );

  onMounted(async () => {
    if (!products.value.length) {
      product.value = await getProduct();

      if (!product.value?.id) {
        isProductValid.value = false;
        return;
      }
    } else {
      product.value = products.value.find(product => product.id === Number(productId.value));
    }

    nextTick(async () => {
      if (product.value?.category) {
        categoryProducts.value = await productsStore.fetchProductsByCategory(true, product.value.category);
      }
    });
    
  });
</script>
