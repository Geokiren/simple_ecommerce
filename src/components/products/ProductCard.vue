<template>
  <div
    class="relative overflow-hidden rounded-lg shadow-lg transition hover:scale-[1.03] duration-300 ease-in-out cursor-pointer  group"
    :ref="isLast ? (el) => { lastItemRef = el as HTMLElement } : undefined"
    @click="goToProductPage"
  >
    <div class="aspect-w-2 aspect-h-1">
      <OptimizedImage
        :src="product.thumbnail"
        alt="Product Thumbnail"
        classes="object-cover w-full h-full"
        @imageLoaded="onImageLoaded"
      />
    </div>

    <div
      class="w-full max-h-0 group-hover:max-h-full opacity-0 group-hover:opacity-90 overflow-hidden transition-all duration-700 ease-in-out absolute left-0 right-0 bottom-0 rounded-b-md bg-secondary  p-4 "
    >
      <h3 class="text-lg font-boldmb-2 text-center">{{ product.title }}</h3>
      <p class="text-sm p-1">{{ product.description }}</p>

      <div class="flex items-center justify-between">
        <div class="flex gap-1 items-center">
          <span>{{ product?.rating }}</span>
          <IconStar classes="size-5" />
        </div>

        <div class="text-xl font-semibold text-right w-full">{{ product.price }}â‚¬</div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
  import { defineProps, ref, onUnmounted, watch } from 'vue';
  import { useRouter } from 'vue-router';
  import { useIntersectionObserver } from '@vueuse/core';
  import { useProductsStore } from '../../stores/products';
  import { storeToRefs } from 'pinia';
  
  import type { Product } from '../../types';

  import OptimizedImage from '../shared/OptimizedImage.vue';
  import IconStar from '../icons/IconStar.vue';

  const { product, isLast } = defineProps<{
    product: Product;
    isLast: boolean;
  }>();

  const productsStore = useProductsStore();
  const { loading } = storeToRefs(productsStore);

  const lastItemRef = ref<HTMLElement | null>(null)
  
  const observerOptions = {
    rootMargin: '300px',
    threshold: 0,
  };

  const stopObserver = ref(() => {});

  const setupObserver = () => {
    stopObserver.value()
    const { stop } = useIntersectionObserver(
      lastItemRef,
      ([entry]) => {
        if (entry?.isIntersecting) {
          stopObserver.value();
          productsStore.incrementPage();
        }
      },
      observerOptions
    )
    stopObserver.value = stop
  }

  onUnmounted(() => {
    stopObserver.value()
  });

  watch(() => isLast, (value) => {
    if (!value) stopObserver.value();
  });

  const onImageLoaded = () => {
    if (isLast && !loading.value) setupObserver();
  };
  
  const router = useRouter();

  const goToProductPage = () => {
    router.push(`/product/${product.id}`);
  };
</script>
