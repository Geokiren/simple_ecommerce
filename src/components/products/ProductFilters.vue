<template>
  <section class="flex flex-col gap-4 items-center justify-between max-w-screen-2xl mx-auto px-6 pt-6 pb-0">
    <input class="w-full bg-inherit focus:border-b px-2 focus:outline-none text-2xl text-center" type="text" v-model="serchQuery" placeholder="Search">

    <div class="w-full flex flex-col lg:flex-row gap-2">
      <div class="flex flex-col w-full gap-1">
        <label class="text-xs font-semibold ps-1">Category</label>
        <DropDown
          v-model="selectedCategory"
          :options="categories"
          label="name"
          value="slug"
          :disabled="!!searchTerm"
          placeholder="Select a category"
        />
      </div>

      <div class="flex flex-col w-full gap-1">
        <label class="text-xs font-semibold ps-1">Sort By</label>
        <DropDown
          :modelValue="sortValue"
          :options="sortingAttributes"
          :hasClearButton="false"
          label="label"
          value="value"
          placeholder="Sort by"
          :disabled="!!searchTerm"
          @update:modelValue="onChangeSort"
        />
      </div>

      <div class="flex flex-col w-full gap-1">
        <label class="text-xs font-semibold ps-1">Order By</label>
        <DropDown
          :modelValue="orderValue"
          :options="sortingDirection"
          :hasClearButton="false"
          label="label"
          value="value"
          placeholder="Order by"
          :disabled="!!searchTerm"
          @update:modelValue="onChangeOrder"
        />
      </div>
    </div>
  </section>
</template>

<script setup lang="ts">
  import { ref, watch, computed } from 'vue';
  import { useProductsStore } from '@/stores/products';
  import { storeToRefs } from 'pinia';

  import DropDown from '../shared/DropDown.vue';

  import type { Category } from '@/types/category'; 

  const productsStore = useProductsStore();

  const { categories, activeCategory, searchTerm } = storeToRefs(productsStore);

  const sortingAttributes = computed(() => {
    return [
      { label: 'Title', value: 'title' },
      { label: 'Price', value: 'price' }
    ]
  });

  const sortingDirection = computed(() => {
    return [
      { label: 'Ascending', value: 'asc' },
      { label: 'Descending', value: 'desc' }
    ]
  });

  const serchQuery = ref<string>('');
  const selectedCategory = ref<Category | null>(activeCategory.value);
  const sortBy = ref<string>('title');
  const order = ref<string>('asc');

  const sortValue = computed(() => {
    return sortingAttributes.value.find((item) => item.value === sortBy.value);
  });

  const orderValue = computed(() => {
    return sortingDirection.value.find((item) => item.value === order.value);
  });

  const onChangeOrder = (selected: { label: string; value: string }) => {
    order.value = selected?.value || 'asc';
  };

  const onChangeSort = (selected: { label: string; value: string }) => {
    sortBy.value = selected?.value || 'title';
  };

  watch(serchQuery, (value) => {
    productsStore.setSearchTerm(value);
  });

  watch(selectedCategory, (value) => {
    productsStore.setActiveCategory(value);
  });

  watch(sortBy, (value) => {
    productsStore.setSorting(value, 'sortBy');
  });

  watch(order, (value) => {
    productsStore.setSorting(value, 'order');
  });
</script>