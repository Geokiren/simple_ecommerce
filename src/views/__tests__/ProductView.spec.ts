import { mount } from '@vue/test-utils';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { nextTick } from 'vue';
import { useRoute } from 'vue-router';
import { useProductsStore } from '@/stores/products';
import { useCartStore } from '@/stores/cart';
import { createPinia } from 'pinia';

import ProductView from '../ProductView.vue';
import InvalidView from '../InvalidView.vue';
import ProductCard from '@/components/products/ProductCard.vue';

import { products } from '../../mockData';

vi.mock('vue-router', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...(actual as Record<string, any>),
    useRoute: vi.fn(),
    useRouter: () => ({
      // mock router properties and methods here
    })
  };
});

describe('ProductView', () => {
  let mockRoute: any;

  useProductsStore(createPinia());
  useCartStore(createPinia());

  beforeEach(() => {
    mockRoute = {
      params: {
        id: '1'
      }
    };
    (useRoute as any).mockReturnValue(mockRoute);
  });

  it('renders product details when product is valid', async () => {
    const wrapper = mount(ProductView);

    const instance = wrapper.vm as any;
    instance.product = products[0];

    await nextTick();

    const title = wrapper.find('[data-test="product-title"]');
    const description = wrapper.find('[data-test="product-description"]');
    const price = wrapper.find('[data-test="product-price"]');
    const discount = wrapper.find('[data-test="product-discount"]');
    const rating = wrapper.find('[data-test="product-rating"]');

    expect(title.text()).toBe('Product 1');
    expect(description.text()).toBe('Product 1 description');
    expect(price.text()).toBe('â‚¬10.99');
    expect(discount.text()).toBe('10%');
    expect(rating.text()).toBe('4.5');
  });

  it('renders InvalidView if product is not valid', async () => {
    const wrapper = mount(ProductView);

    const instance = wrapper.vm as any;
    instance.isProductValid = false;

    await nextTick();

    expect(wrapper.findComponent(InvalidView).exists()).toBe(true);
    expect(wrapper.find('[data-test="product-view"]').exists()).toBe(false);
  });

  it('renders reviews correctly', async () => {
    const wrapper = mount(ProductView);

    const instance = wrapper.vm as any;
    instance.product = products[0];

    await nextTick();

    const reviewElements = wrapper.findAll('[data-test="review-name"]');

    expect(reviewElements.length).toBe(1);
    expect(reviewElements[0].text()).toBe('John Doe');
    expect(wrapper.findAll('[data-test="review-comment"]')[0].text()).toBe('Great product!');
  });

  it('renders related category products except current one', async () => {
    const wrapper = mount(ProductView);

    const instance = wrapper.vm as any;
    instance.categoryProducts.push(products[1]);

    await nextTick();

    expect(wrapper.findAllComponents(ProductCard)).toHaveLength(1);
    expect(wrapper.findAllComponents(ProductCard)[0].props('product').title).toBe('Product 2');
  });

  it('matches snapshot', async () => {
    const wrapper = mount(ProductView);

    const instance = wrapper.vm as any;
    instance.product = products[0];
    instance.categoryProducts.push(products[1]);

    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });
});
