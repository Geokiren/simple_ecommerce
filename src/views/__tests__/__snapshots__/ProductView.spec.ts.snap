// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductView > matches snapshot 1`] = `
<section
  class="max-w-screen-2xl mx-auto px-6 py-6 flex flex-col gap-4"
  data-test="product-view"
>
  <div
    class="flex flex-col lg:flex-row"
  >
    <div
      class="w-full lg:w-2/3 h-auto lg:pr-2"
    >
      <!--v-if-->
    </div>
    <div
      class="w-full lg:w-1/3 mt-6 lg:mt-0 lg:ps-2 flex flex-col gap-2 lg:gap-10"
    >
      <h1
        class="text-xl text-center font-bold mb-2"
        data-test="product-title"
      >
        Product 1
      </h1>
      <div
        class="flex justify-between items-center"
      >
        <div
          class="flex gap-1 flex-wrap"
        >
          
          
        </div>
        <div
          class="flex gap-1 items-center"
        >
          <span
            data-test="product-rating"
          >
            4.5
          </span>
          <svg
            class="size-5"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.788 3.21c.448-1.077 1.976-1.077 2.424 0l2.082 5.006 5.404.434c1.164.093 1.636 1.545.749 2.305l-4.117 3.527 1.257 5.273c.271 1.136-.964 2.033-1.96 1.425L12 18.354 7.373 21.18c-.996.608-2.231-.29-1.96-1.425l1.257-5.273-4.117-3.527c-.887-.76-.415-2.212.749-2.305l5.404-.434 2.082-5.005Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <p
        class="text-sm leading-6 font-light"
        data-test="product-description"
      >
        Product 1 description
      </p>
      <div
        class="flex"
      >
        <div
          class="p-2 bg-primary text-xl font-semibold rounded-s text-alttext dark:text-text"
          data-test="product-discount"
        >
          10%
        </div>
        <div
          class="text-xl font-semibold rounded p-2 bg-secondary rounded-e"
          data-test="product-price"
        >
          €10.99
        </div>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <button
          class="btn bg-primary rounded p-4 text-lg hover:bg-active dark:hover:text-alttext transition duration-300 ease-in-out text-alttext dark:text-text mt-4 lg:mt-0"
          data-test="product-add-to-bag"
        >
           ADD TO BAG 
        </button>
        <div
          class="text-xs lg:text-sm text-center"
        >
          <span
            data-test="product-return-policy"
          />
          <span>
             | 
          </span>
          <span
            data-test="product-shipping-info"
          />
          <span>
             | 
          </span>
          <span
            data-test="product-warranty-info"
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <h1
      class="text-xl text-center lg:text-left font-bold px-4 pt-4 pb-2 w-full border-b"
    >
      Products in category
    </h1>
    <transition-group-stub
      appear="false"
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 p-6 max-w-screen-2xl mx-auto gap-3"
      css="true"
      persisted="false"
      tag="section"
    >
      
      
      <div
        class="relative overflow-hidden rounded-lg shadow-lg transition hover:scale-[1.03] duration-300 ease-in-out cursor-pointer group"
      >
        <div
          class="aspect-w-2 aspect-h-1"
        >
          <div
            class="relative overflow-hidden object-cover w-full h-full"
          >
            <img
              alt="Product Thumbnail"
              class="inset-0 w-full h-full object-cover transition-opacity duration-300 opacity-0"
              data-src=""
              src=""
            />
            <!--v-if-->
            <!--v-if-->
          </div>
        </div>
        <div
          class="w-full max-h-0 group-hover:max-h-full opacity-0 group-hover:opacity-90 overflow-hidden transition-all duration-700 ease-in-out absolute left-0 right-0 bottom-0 rounded-b-md bg-secondary p-4"
        >
          <h3
            class="text-lg font-boldmb-2 text-center"
          >
            Product 2
          </h3>
          <p
            class="text-sm p-1"
          />
          <div
            class="flex items-center justify-between"
          >
            <div
              class="flex gap-1 items-center"
            >
              <span>
                0
              </span>
              <svg
                class="size-5"
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10.788 3.21c.448-1.077 1.976-1.077 2.424 0l2.082 5.006 5.404.434c1.164.093 1.636 1.545.749 2.305l-4.117 3.527 1.257 5.273c.271 1.136-.964 2.033-1.96 1.425L12 18.354 7.373 21.18c-.996.608-2.231-.29-1.96-1.425l1.257-5.273-4.117-3.527c-.887-.76-.415-2.212.749-2.305l5.404-.434 2.082-5.005Z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="text-xl font-semibold text-right w-full"
            >
              20.49€
            </div>
          </div>
        </div>
      </div>
      
      
    </transition-group-stub>
  </div>
  <div
    class="flex flex-col gap-2 lg:gap-4"
  >
    <h1
      class="text-xl text-center lg:text-left font-bold px-4 pt-4 pb-2 w-full border-b"
    >
      Reviews
    </h1>
    <div
      class="grid grid-cols-1 lg:grid-cols-3 gap-4"
    >
      
      <div
        class="p-4 bg-secondary rounded-lg shadow-md"
      >
        <div
          class="flex items-center justify-between"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="w-12 h-12 rounded-full overflow-hidden"
            >
              <div
                class="rounded-full w-full h-full bg-primary text-alttext dark:text-text font-bold text-lg flex items-center justify-center"
              >
                <div
                  data-test="review-initials"
                >
                  JD
                </div>
              </div>
            </div>
            <div
              class="text-lg font-bold"
              data-test="review-name"
            >
              John Doe
            </div>
          </div>
          <div
            class="flex items-center gap-1"
          >
            <span
              class="text-lg"
              data-test="review-rating"
            >
              5
            </span>
            <svg
              class="size-5"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10.788 3.21c.448-1.077 1.976-1.077 2.424 0l2.082 5.006 5.404.434c1.164.093 1.636 1.545.749 2.305l-4.117 3.527 1.257 5.273c.271 1.136-.964 2.033-1.96 1.425L12 18.354 7.373 21.18c-.996.608-2.231-.29-1.96-1.425l1.257-5.273-4.117-3.527c-.887-.76-.415-2.212.749-2.305l5.404-.434 2.082-5.005Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="mt-2 text-lg"
        >
          <p
            data-test="review-comment"
          >
            Great product!
          </p>
        </div>
      </div>
      
    </div>
  </div>
</section>
`;
